pallets/stream-payment/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+        type AssetId: Debug + Clone + FullCodec + TypeInfo + MaxEncodedLen + PartialEq + Eq;
+
+        /// Provide interaction with assets.
+        type Assets: Assets<Self::AccountId, Self::AssetId, Self::Balance>;
+
+        /// Represents which units of time can be used. Designed to be an enum
+        /// with a variant for each kind of time source/scale supported.
+        type TimeUnit: Debug + Clone + FullCodec + TypeInfo + MaxEncodedLen + Eq;
+
+        /// Provide the current time in given unit.
+        type TimeProvider: TimeProvider<Self::TimeUnit, Self::Balance>;
+    }
+
+    type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+    type AssetIdOf<T> = <T as Config>::AssetId;
+
+    /// A stream payment from source to target.
+    /// Stores the last time the stream was updated, which allows to compute
+    /// elapsed time and perform payment.
+    #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+    #[derive(RuntimeDebug, PartialEq, Eq, Encode, Decode, Clone, TypeInfo)]
+    pub struct Stream<AccountId, Unit, AssetId, Balance> {
+        pub source: AccountId,
+        pub target: AccountId,
+        pub time_unit: Unit,
+        pub asset_id: AssetId,
+        pub rate_per_time_unit: Balance,
+        pub deposit: Balance,
+        pub last_time_updated: Balance,
+    }
+
+    pub type StreamOf<T> =
+        Stream<AccountIdOf<T>, <T as Config>::TimeUnit, AssetIdOf<T>, <T as Config>::Balance>;
+
+    /// Store the next available stream id.
+    #[pallet::storage]
+    pub type NextStreamId<T: Config> = StorageValue<Value = T::StreamId, QueryKind = ValueQuery>;
+
+    /// Store each stream indexed by an Id.
+    #[pallet::storage]
+    pub type Streams<T: Config> = StorageMap<
+        Hasher = Blake2_128Concat,
+        Key = T::StreamId,
+        Value = StreamOf<T>,
+        QueryKind = OptionQuery,
+    >;
+
+    /// Lookup for all streams with given source.
+    /// To avoid maintaining a growing list of stream ids, they are stored in
+    /// the form of an entry (AccountId, StreamId). If such entry exists then
+    /// this AccountId is a source in StreamId. One can iterate over all storage
+    /// keys starting with the AccountId to find all StreamIds.
+    #[pallet::storage]
+    pub type LookupStreamsWithSource<T: Config> = StorageDoubleMap<
+        Key1 = AccountIdOf<T>,
+        Hasher1 = Blake2_128Concat,
+        Key2 = T::StreamId,
+        Hasher2 = Blake2_128Concat,
+        Value = (),
+        QueryKind = OptionQuery,
+    >;
+
+    /// Lookup for all streams with given target.
+    /// To avoid maintaining a growing list of stream ids, they are stored in
+    /// the form of an entry (AccountId, StreamId). If such entry exists then
+    /// this AccountId is a target in StreamId. One can iterate over all storage
+    /// keys starting with the AccountId to find all StreamIds.
+    #[pallet::storage]
+    pub type LookupStreamsWithTarget<T: Config> = StorageDoubleMap<
```

Comment:
These double maps are not used by the pallet, it would probably be better to remove them and calculate them offchain. Unless they should be used somewhere?

Reply:
They are here for the UI to find the list of streams one is the source/target.